### DISCUSSION TITLE
Connecting / Disconnecting events

### DISCUSSION THREAD

**Discussion by b00f** [libp2p/rust-libp2p#1984] [Q&A]
Is there any way to know if a peer or node is connected (or disconnected) from the network?
I checked the swarm code but I couldn't find a general way.

**Comment by mxinden** [comment]
When creating a swarm you have to provide a structure implementing the [`NetworkBehaviour` trait](https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/behaviour.rs). This structure is notified of connection establishment / connection termination through the `inject_connected` and `inject_disconnected` methods.

**Comment by b00f** [comment]
@mxinden  Imagine I have a Behaviour like this:
``` rust
#[derive(NetworkBehaviour)]
#[behaviour(out_event = "BehaviourEventOut", poll_method = "poll")]
pub struct Behaviour {
    gossipsub: Gossipsub,
    mdns: Toggle<Mdns>,
    ping: Ping,
    identify: Identify,
    // kademlia: ...,
    // bitswap: Bitswap,
}
```
As you can see I have used NetworkBehaviour drive macro . However the inject_connected is inside the NetworkBehaviour trait. Shall I re-implement all the NetworkBehaviour methods and not use the derived macro?
Any idea or example?

**Comment by koivunej** [comment]
@b00f your setup looks quite bit similar to what we have in [rust-ipfs](https://github.com/rs-ipfs/rust-ipfs). I'd recommend you to add a new component to your Behaviour which would implement the NetworkBehaviour trait. If you are after accounting (which peers are connected etc) on demand, the [`ipfs::p2p::SwarmApi`](https://github.com/rs-ipfs/rust-ipfs/blob/1fe59bd777a25cb658185df0d4154be42333d701/src/p2p/swarm.rs#L133-L253) might be an example. There might be parts of data we store for no good reason (this could be queried through [`Swarm::connection_info`](https://docs.rs/libp2p/0.28.1/libp2p/swarm/struct.ExpandedSwarm.html#method.connection_info) but maybe we weren't aware of the `connection_info` when implementing this in the first place), and likely you do not need any of the subscription stuff we have.

**Comment by b00f** [comment]
@koivunej Thanks for the explanation. Not an easy way as I expected before but it works as you said.
 
I have created a boilerplate for libp2p [here](https://github.com/b00f/libp2p_rust_boilerplate)

First of all, I really don't like my implementation for events. As you can see I have duplicated codes, [here](https://github.com/b00f/libp2p_rust_boilerplate/blob/54a4c3f997aaa4df3c0e9589bf6398222a02a148/network/src/swarm_api.rs#L11), [here](https://github.com/b00f/libp2p_rust_boilerplate/blob/54a4c3f997aaa4df3c0e9589bf6398222a02a148/network/src/event.rs#L4) and [here](https://github.com/b00f/libp2p_rust_boilerplate/blob/54a4c3f997aaa4df3c0e9589bf6398222a02a148/network/src/behaviour.rs#L52) . Any suggestion to make it better?

I also noticed that sometimes some events are missed. If you run the boilerplate you can see that `Hi` and `Goodbye` messages are randomly missed.  

Another problem is that broadcasting immediately after PeerConnected event [doesn't work properly](https://github.com/b00f/libp2p_rust_boilerplate/issues/1). Any idea?
